apm-sniffer/apm-sdk-plugin/pulsar-plugin/src/main/java/org/apache/skywalking/apm/plugin/pulsar/PulsarConsumerListenerInterceptor.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.pulsar;
+
+import org.apache.pulsar.client.api.Message;
+import org.apache.skywalking.apm.agent.core.context.CarrierItem;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.tag.Tags;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceMethodsAroundInterceptor;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+
+import java.lang.reflect.Method;
+
+/**
+ * Interceptor for pulsar consumer message listener enhanced instance
+ * <p>
+ * Here is the intercept process steps:
+ *
+ * <pre>
+ *  1. Get @{@link ContextCarrier} from message
+ *  2. Create a local span when call <code>received</code> method
+ *  3. Extract trace information from context carrier
+ *  4. Stop the local span when <code>received</code> method finished.
+ * </pre>
+ */
+public class PulsarConsumerListenerInterceptor implements InstanceMethodsAroundInterceptor {
+
+    public static final String CONSUMER_OPERATE_NAME = "Pulsar/Consumer/MessageListener";
```

Comment:
Shall we need to introduce a new operation name? The message listener just a way to consume messages, for the tracing context, we just need to know the consumer got the message.

And, the message listener can't work together with the receive() method, so I think we can use the existing operation name.

Reply:
Before entering the `MessageListener` callback method, the `receive()` method will be called to receive the message first, so I think introduce a new operation name can distinguish these two spans

Reply:
It's not the same concepts here I think, the `Pulsar/Producer` means the client sends out the message, the `Pulsar/Producer/Callback` means the broker sends out the publish receipt to the client, if not callback from the broker, this means the message does not written succeed. 

But for the message listener, it's a way at the client side to get messages, both `Pulsar/Consumer/MessageListener` and `Pulsar/Consumer` are indicate the messages are arrivals the consumer, and we are not support the `Pulsar/Consumer/Acknowledge` for now to indicate the consumer processed the message and acked it.

The reason that I think we don't need a new operation here is we might have different get message methods in the client side such as batchReceive etc., I think the trace is telling users that the message arrives the consumer is enough here, If we want to add more details about how the users get the messages, I think we can add some labels in the trace context, but not a new operation name.

BTW, do we need to add a new span for handling this case? Looks like

```
--- Pulsar/Consumer
      --- Pulsar/Consumer/MessageListener
```
If there is a way to set the existing span to the message listener thread, seems more reasonable here.

Reply:
I know what you mean, but in most cases, we will write the business code directly in the `MessageListener`, therefore it is not a span of message reception, but more like a process of message handling, such as
```
// span Pulsar/Consumer start
Message message = consumer.receive();
// List<Message> messages = consumer.batchReceive();
// span Pulsar/Consumer end

// business code, maybe start a new span named HandleMessage
// do some business  logic with message...
consumer.acknowledge(message);
// business code end, and the HandleMessage span is end
```

And with `MessageListener`, we will do the business logic like this
```
pulsarClient.newConsumer().topic("test").subscriptionName("test").messageListener((consumer,message) -> {
    // span Pulsar/Consumer/MessageListener start
    // do some business logic with message...
    consumer.acknowledge(message);
    // span Pulsar/Consumer/MessageListener end
})
```
Span `Pulsar/Consumer` is finished in client internal, and the `Pulsar/Consumer/MessageListener` is a `Local` span that represent the business logic processing. So I agree with adding a new span with `Pulsar/Consumer` span as parent.